### ISSUE TITLE
[Performance]  ONNXRuntime 1.14.0 segfaults during inference

### ISSUE THREAD

**Issue by vahvero** [microsoft/onnxruntime#14691]
### Describe the issue

Torch exported mask rCNN model segfaults during inference in 1.14.0, but passes in 1.13.1.

Used packages:

```
name         : torch                                                                      
 version      : 1.12.1                                                                     
 description  : Tensors and Dynamic neural networks in Python with strong GPU acceleration 

dependencies
 - typing-extensions *

required by
 - torchmetrics >=1.8.1
 - torchvision 1.12.1
```
```
name         : torchvision                                                 
 version      : 0.13.1                                                      
 description  : image and video datasets and models for torch deep learning 

dependencies
 - numpy *
 - pillow >=5.3.0,<8.3.0 || >=8.4.0
 - requests *
 - torch 1.12.1
 - typing-extensions *
```

```
 name         : onnxruntime-gpu                                                   
 version      : 1.14.0                                                            
 description  : ONNX Runtime is a runtime accelerator for Machine Learning models 

dependencies
 - coloredlogs *
 - flatbuffers *
 - numpy >=1.21.6
 - packaging *
 - protobuf *
 - sympy *
```

Result of the reproducing script

```
<redacted>
2023-02-15 11:40:19.986223786 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '4465'. It is not used by any node and should be removed from the model.
2023-02-15 11:40:19.986273651 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '4762'. It is not used by any node and should be removed from the model.
2023-02-15 11:40:19.986279042 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '4763'. It is not used by any node and should be removed from the model.
2023-02-15 11:40:19.986285863 [W:onnxruntime:, graph.cc:3490 CleanUnusedInitializersAndNodeArgs] Removing initializer '4765'. It is not used by any node and should be removed from the model.
Segmentation fault (core dumped)
```

### To reproduce

```
# %% Imports

import torch
from torchvision.models.detection.mask_rcnn import (
    maskrcnn_resnet50_fpn,
    MaskRCNN_ResNet50_FPN_Weights,
)
import onnxruntime as ort


# %% Create model
mask_rcnn = maskrcnn_resnet50_fpn(weights=MaskRCNN_ResNet50_FPN_Weights.DEFAULT)

example_input = torch.rand(2, 3, 1028, 1028)
filename = "/tmp/mask_rcnn.onnx"

# %% Run export
torch.onnx.export(
    mask_rcnn,
    example_input,
    filename,
    opset_version=16,
)

# %% Attempt running

session = ort.InferenceSession(
    filename,
    providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
)
model_inputs = session.get_inputs()[0]

batch = {
    model_inputs.name: example_input.numpy(),
}

# %% Run
for _ in range(100):
    resp = session.run(None, batch)

# %%
```

### Urgency

High, package is unusable or at least unstable

### Platform

Linux

### OS Version

Ubuntu 20.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.14.0

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

 NVIDIA-SMI 520.61.05    Driver Version: 520.61.05    CUDA Version: 11.8   

### Model File

File is created in reproducing script

### Is this a quantized model?

Unknown

**Comment by stevenlix** [comment]
Thanks for reporting the issue. I've reproduced the issue. Will take a further look.

**Comment by satyajandhyala** [comment]
The RoiAlign opset 16 is not supported on CPU EP currently. The fall back CPU EP appears to have an issue. If the loop is run only once the bug is not reproducible. If session.run is called twice then then bug is reproducible. For opset 16 the following additional operators (compared to opset 15) are assigned to CPU EP
Flatten
Identity
If
Loop
RoiAlign
ScatterElements
Shape

**Comment by Callidior** [comment]
I think I encountered the same or at least a similar issue and wanted to share my code, since it is small and easy to reproduce.

I export the following PyTorch model to ONNX:

```python
import torch
import torch.nn.functional as F
from torch import nn, Tensor


@torch.jit.script_if_tracing
def nonzero_coords_padded(heatmap: Tensor) -> Tensor:

    # Find nonzero points
    points = [torch.nonzero(m) for m in heatmap]

    # Create output tensors
    max_num = 0
    for c in points:
        max_num = max(max_num, len(c))
    det_points: Tensor = heatmap.new_full((len(points), max_num, 2), -1, dtype=torch.long)

    for i, c in enumerate(points):
        det_points[i, :c.size(0)] = c

    return det_points


class Model(nn.Module):
    def forward(self, x: Tensor) -> Tensor:
        return nonzero_coords_padded(x)
```

When I run the ONNX model, I don't get a segfault but sometimes incorrect results. Even with the same input data, the result is sometimes correct, sometimes wrong. This can be worked around by downgrading ONNX runtime to <= 1.13 or exporting to opset < 16.

My test input is created as follows:

```python
import numpy as np

true_centers = [
    [[127, 68], [99, 245]],
    [[11, 197], [222, 33], [123, 231]]
]
test_inputs = np.zeros((2, 256, 256), dtype=np.float32)
for i, tc in enumerate(true_centers):
    test_inputs[i][tuple(np.asarray(tc).T)] = 1
```

The correct model output looks as follows:

```python
array([[[ 99, 245],
        [127,  68],
        [ -1,  -1]],

       [[ 11, 197],
        [123, 231],
        [222,  33]]], dtype=int64)
```

About 25% of the time, I get the following output instead:

```python
array([[[ 11, 197],
        [123, 231],
        [ -1,  -1]],

       [[ 11, 197],
        [123, 231],
        [222,  33]]], dtype=int64)
```

Here, the difference is small. If my model and input data grow bigger, the output becomes complete rubbish.

I suspect that the for loop is part of the issue.

**Comment by RandySheriffH** [comment]
@vahvero: the fix's been merged - mind try building from the latest main for an update?

**Comment by vahvero** [comment]
I came up with some dependency issues when building from source. I can probably try with a deeper dive again later during the weekend.

**Comment by RandySheriffH** [comment]
Thanks for the confirmation!
Closing the issue but feel free to reach out for further assistance.
